### ISSUE TITLE
Module doesn't start when the iotedge runtime cannot pull it, even though the latest version is already present

### ISSUE THREAD

**Issue by FMauroy** [Azure/iotedge#4552]
## Expected Behavior
Due to a network config issue at a customer site, the runtime was not able to check for the latest version of a custom module (failed with an authentication error, even though the deployment has been working for months). It's a pity that it stops there and doesn't try to run the version that is already present on the system (exact same address and tag). As this is a production environment, we'd rather run an older version of the module than not at all!
Note: docker pull worked, why the runtime could not perform the same action is still a mystery.

==> expected behaviour: runtime fails to check/update the latest image, sees that a version with the same name and tag already exists, reports a warning and loads the existing image. It may retry the update process at a later time.

### Output of `iotedge check`
no error in check (only a warning about not finding an NTP server)

### Device Information
* Host OS: Windows 10 IoT 17723
* Architecture: amd64
* Container OS: Windows containers

### Runtime Versions
* aziot-edged:  iotedge 1.1.0 92ac4878a4c8ca74c0db14cd2e63273f1c8661f4 (noticed the problem on 1.0.10.4 then upgraded to 1.1.0, no change)
* Edge Agent: 1.1.0
* Edge Hub: 1.1.0
* Docker/Moby: Moby -> Docker version 19.03.12+azure, build 0ed913b885c8919944a2e4c8d0b80a318a8dd48b

## Logs

<details>
<summary>edge-agent logs</summary>

2021-03-05 17:29:10.885 +02:00 [WRN] - Reconcile failed because of the an exception
System.AggregateException: One or more errors occurred. (Error calling Create module REDACTED: Could not create module REDACTED
        caused by: Could not pull image REDACTED.azurecr.io/REDACTED:0.1
        caused by: unauthorized: authentication required)
 ---> Microsoft.Azure.Devices.Edge.Agent.Edgelet.EdgeletCommunicationException- Message:Error calling Create module REDACTED: Could not create module REDACTED
        caused by: Could not pull image REDACTED.azurecr.io/REDACTED:0.1
        caused by: unauthorized: authentication required, StatusCode:500, at:   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2020_07_07.ModuleManagementHttpClient.HandleException(Exception exception, String operation) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\version_2020_07_07\ModuleManagementHttpClient.cs:line 235
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\versioning\ModuleManagementHttpClientVersioned.cs:line 146
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2020_07_07.ModuleManagementHttpClient.CreateModuleAsync(ModuleSpec moduleSpec) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\version_2020_07_07\ModuleManagementHttpClient.cs:line 102
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\LoggingCommandFactory.cs:line 67
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\commands\GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\LoggingCommandFactory.cs:line 67
   at Microsoft.Azure.Devices.Edge.Agent.Core.PlanRunners.OrderedRetryPlanRunner.ExecuteAsync(Int64 deploymentId, Plan plan, CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\planrunners\OrdererdRetryPlanRunner.cs:line 87
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Edge.Agent.Core.PlanRunners.OrderedRetryPlanRunner.<>c.<ExecuteAsync>b__7_0(List`1 f) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\planrunners\OrdererdRetryPlanRunner.cs:line 115
   at Microsoft.Azure.Devices.Edge.Agent.Core.PlanRunners.OrderedRetryPlanRunner.ExecuteAsync(Int64 deploymentId, Plan plan, CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\planrunners\OrdererdRetryPlanRunner.cs:line 116
   at Microsoft.Azure.Devices.Edge.Agent.Core.Agent.ReconcileAsync(CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\Agent.cs:line 203
   at Microsoft.Azure.Devices.Edge.Agent.Core.Agent.ReconcileAsync(CancellationToken token) in C:\agent\_work\9\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Core\Agent.cs:line 203

</details>

## Additional Information

After the network team investigated the problem, they found an issue with the proxy. This caused security popups to show up in the browser, and apparently caused the authentication errors in the runtime. In normal times, the edge devices are in a list of authorized machines and do not require authentication or use of the proxy.

**Comment by and-rewsmith** [comment]
@FMauroy 
> Note: docker pull worked, why the runtime could not perform the same action is still a mystery.

It sounds like you figured out that the runtime auth errors were caused by your proxy configuration?

If you don't care about updates to your module you can turn the imagePullPolicy off for edge agent as described here to mitigate:
https://docs.microsoft.com/en-us/azure/iot-edge/module-edgeagent-edgehub?view=iotedge-2018-06#edgeagent-desired-properties

You are right though in that we shouldn't block the module from running if we can't update it. Please note though that I believe this will only be a problem if either the iotedge service restarts or a deployment update is triggered. So hopefully if the module has started it will continue running until either of these events are triggered by your architecture. Are you seeing this problem happen outside of these events?

**Comment by FMauroy** [comment]
Thanks for your quick reply.

I reported this issue as I think this is an edge case (no pun) that you might have missed. At least, this is the first time we saw this behaviour.
I'll change the setting you mention to never update the module. I suppose it will still update when we change the image tag? When we deploy to production, we make the CI/CD pipeline in Azure DevOps tag the new image with an incremental number.

In a nutshell, I see 2 points of interest:
- "docker pull" worked, even with the wrongly configured proxy. Does the edge runtime pull images differently?
- if the image is already on the system, a failure to update it should be ignored and reported as a warning, and the existing image loaded anyway;

That would be a nice improvement, with little changes to the code, and improve the resiliency of the system.
Cheers

**Comment by chipalost** [comment]
Removing the no-issue-activity label as this is an issue we're still looking at internally.

**Comment by varunpuranik** [comment]
@FMauroy - I would like to understand your scenario a bit more. 

The IoT Edge deployment specified as part of the EdgeAgent twin allows setting the [value for the flag imagePullPolicy](https://docs.microsoft.com/en-us/azure/iot-edge/module-edgeagent-edgehub?view=iotedge-2020-11#edgeagent-desired-properties), which has 2 values - 
- onCreate (default): If this value is specified, then the IoT Edge agent will try to pull the image before creating the module, and fail if the module pull fails. But if this is an upgrade scenario (say upgrading version), then the existing module will not be removed. 
- never: If this value is specified, then the IoT Edge agent will not try to pull the image, and instead expect the image to be present on the disk and create the module from that image. 

So, if your scenario is for creating a new module, then you can use the flag imagePullPolicy=never. If it is an upgrade scenario, then IoT Edge should not have to create a new module, as the existing module should continue to run if the pull of the new container image fails.

**Comment by varunpuranik** [comment]
@FMauroy - we are trying to understand the scenario where the flags IoT Edge currently has don't work. 
Since we have not heard back from you, I will close this issue. Feel free to re-open it if you have more details/context for us.

**Comment by Taras-Tyrsa** [comment]
I've experienced the same issue. Specific version of module is already present on device. Even though it is present, Edge Agent tries to pull image but fails due to some reason (in my case, outdated credentials to container registry) and module not started.

In this scenario it should just give warning and start module anyway, I think...
